Sources/PackageGraph/DependencyResolver.swift

```diff
@@ -171,21 +171,51 @@ public protocol PackageContainerProvider {
 public struct PackageContainerConstraint<T: PackageContainerIdentifier>: CustomStringConvertible {
     public typealias Identifier = T
 
+    public enum Requirement: Equatable {
+        case versionSet(VersionSetSpecifier)
+        case unversioned
+
+        public static func ==(lhs: Requirement, rhs: Requirement) -> Bool {
+            switch (lhs, rhs) {
+            case (.unversioned, .unversioned):
+                return true
+            case (.unversioned, _):
+                return false
+            case (.versionSet(let lhs), .versionSet(let rhs)):
+                return lhs == rhs
+            case (.versionSet, _):
+                return false
+            }
+        }
+    }
+
     /// The identifier for the container the constraint is on.
     public let identifier: Identifier
 
     /// The version requirements.
-    public let versionRequirement: VersionSetSpecifier
+    public let requirement: Requirement
+
+    public var versionRequirement: VersionSetSpecifier? {
```

Comment:
Do we need this property, versus just handling the cases directly at each site?

IIRC, it was already a little weird that we had combined "version" with "requirement" here, so it might make the code more readable to ditch this.